chainstate/test-suite/src/tests/delegation_tests.rs

```diff
@@ -0,0 +1,458 @@
+        let tx = TransactionBuilder::new()
+            .add_input(delegate_staking_outpoint.into(), empty_witness(&mut rng))
+            .add_output(TxOutput::SpendShareFromDelegation(
+                spend_change,
+                Destination::AnyoneCanSpend,
+                delegation_id,
+                OutputTimeLock::ForBlockCount(1),
+            ))
+            .build();
+
+        tf.make_block_builder().add_transaction(tx).build_and_process().unwrap();
+
+        let delegation_balance = PoSAccountingStorageRead::<TipStorageTag>::get_delegation_balance(
+            &tf.storage,
+            delegation_id,
+        )
+        .unwrap();
+        assert!(delegation_balance.is_none());
+
+        let delegation_data = PoSAccountingStorageRead::<TipStorageTag>::get_delegation_data(
+            &tf.storage,
+            delegation_id,
+        )
+        .unwrap()
+        .unwrap();
+        assert_eq!(
+            delegation_data,
+            DelegationData::new(pool_id, Destination::AnyoneCanSpend)
+        );
+    });
+}
+
+#[rstest]
+#[trace]
+#[case(Seed::from_entropy())]
+fn delegation_cleanup(#[case] seed: Seed) {
+    utils::concurrency::model(move || {
+        let mut rng = make_seedable_rng(seed);
+        let mut tf = TestFramework::builder(&mut rng).build();
+
+        let (pool_id, stake_outpoint, delegation_id, _, transfer_outpoint) =
+            prepare_delegation(&mut rng, &mut tf);
+        let amount_to_delegate = get_output_value(
+            utxo::UtxosStorageRead::get_utxo(&tf.storage, &transfer_outpoint)
+                .unwrap()
+                .unwrap()
+                .output(),
+        )
+        .unwrap()
+        .coin_amount()
+        .unwrap();
+
+        // Delegate staking
+        let tx = TransactionBuilder::new()
+            .add_input(transfer_outpoint.into(), empty_witness(&mut rng))
+            .add_output(TxOutput::DelegateStaking(amount_to_delegate, delegation_id))
+            .build();
+
+        let delegate_staking_outpoint =
+            OutPoint::new(OutPointSourceId::Transaction(tx.transaction().get_id()), 0);
+        tf.make_block_builder().add_transaction(tx).build_and_process().unwrap();
+
+        let delegation_balance = PoSAccountingStorageRead::<TipStorageTag>::get_delegation_balance(
+            &tf.storage,
+            delegation_id,
+        )
+        .unwrap();
+        assert_eq!(Some(amount_to_delegate), delegation_balance);
+
+        // decommission the pool
+        let tx = TransactionBuilder::new()
+            .add_input(stake_outpoint.into(), empty_witness(&mut rng))
+            .add_output(TxOutput::DecommissionPool(
+                Amount::from_atoms(1),
+                Destination::AnyoneCanSpend,
+                pool_id,
+                OutputTimeLock::ForBlockCount(1),
+            ))
+            .build();
+        tf.make_block_builder().add_transaction(tx).build_and_process().unwrap();
+
+        // Spend all delegation share
```

Comment:
Please do this in two steps at least. I would love seeing a randomized test that spends chunks until it's empty, then try one more time after that and confirm that it fails.

Reply:
added 2 steps for now